autopush_rs/src/server/tls.rs

```diff
@@ -0,0 +1,129 @@
+//! TLS support for the autopush server
+//!
+//! Currently tungstenite in the way we use it just operates generically over an
+//! `AsyncRead`/`AsyncWrite` stream, so this provides a `MaybeTlsStream` type
+//! which dispatches at runtime whether it's a plaintext or TLS stream after a
+//! connection is established.
+
+use std::fs::File;
+use std::io::{self, Read, Write};
+use std::path::Path;
+use std::rc::Rc;
+
+use futures::future;
+use futures::{Poll, Future};
+use openssl::dh::Dh;
+use openssl::pkey::PKey;
+use openssl::ssl::{SslAcceptor, SslAcceptorBuilder, SslMethod};
+use openssl::x509::X509;
+use tokio_core::net::TcpStream;
+use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_openssl::{SslStream, SslAcceptorExt};
+
+use server::{Server, ServerOptions};
+use errors::*;
+
+/// Creates an `SslAcceptor`, if needed, ready to accept TLS connections.
+///
+/// This method is called early on when the server is created and the
+/// `SslAcceptor` type is stored globally in the `Server` structure, later used
+/// to process all accepted TCP sockets.
+pub fn configure(opts: &ServerOptions) -> Option<SslAcceptor> {
+    let key = match opts.ssl_key {
+        Some(ref key) => read(key),
+        None => return None,
+    };
+    let key = PKey::private_key_from_pem(&key)
+        .expect("failed to create private key");
+    let cert = read(opts.ssl_cert.as_ref().expect("ssl_cert not configured"));
+    let cert = X509::from_pem(&cert).expect("failed to create certificate");
+    let dh_param = opts.ssl_dh_param.as_ref().expect("ssl_dh_param not configured");
+    let dh_param = Dh::from_pem(&read(dh_param)).expect("failed to create dh");
+
+    let mut builder = SslAcceptorBuilder::mozilla_intermediate(
```

Comment:
y'all probably know more about the desired options here, so it may be useful to browse the [`SslAcceptorBuilder`](https://docs.rs/openssl/0.9.17/openssl/ssl/struct.SslAcceptorBuilder.html) docs and [`SslContextBuilder`](https://docs.rs/openssl/0.9.17/openssl/ssl/struct.SslContextBuilder.html) docs to see if this is missing anything!